third_party/intel/lib/TritonRaiseBlockPointer/TritonRaiseBlockPointer.cpp

```diff
@@ -0,0 +1,299 @@
+//===----------------------------------------------------------------------===//
+//
+// Copyright (c) Microsoft Corporation, Meta Platforms.
+// Licensed under the MIT license.
+//
+//===----------------------------------------------------------------------===//
+
+#include "intel/include/TritonRaiseBlockPointer/Passes.h"
+
+#include "mlir/IR/Matchers.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "triton-raise-block-pointer"
+
+using namespace mlir;
+
+namespace mlir::triton::intel {
+#define GEN_PASS_DEF_TRITONRAISEBLOCKPOINTER
+#include "intel/include/TritonRaiseBlockPointer/Passes.h.inc"
+} // namespace mlir::triton::intel
+
+namespace {
+constexpr unsigned offsetBitwidth = 32;
+constexpr unsigned shapeAndStridesBitwidth = 64;
+
+// Data structure used to decode pointer arithmetics. offsets, sizes, and
+// strides are in unit of elements in a linearly laid-out memory, which is the
+// same as pointer arithmetic operations in Triton language. scalar is a
+// shortcut used when the entire state describes a single scalar value. source
+// is the base pointer. If order is present, PtrState describes block pointer;
+// otherwise it describes non-block pointers. When it describes block pointer,
+// shape field means the same field as tt.make_tensor_ptr; when it describes a
+// non-block pointer, shape field indicates how address wraps around (i.e.,
+// modulo); a constant 0 indicates no modulo for the dimension.
+struct PtrState {
+
+  SmallVector<Value> offsets;
+  SmallVector<Value> strides;
+  SmallVector<Value> shape;
+  SmallVector<int32_t> sizes;
+  SmallVector<int32_t> order;
+
+  Value source;
+  Value scalar;
+
+  int32_t getRank() const {
+    assert(offsets.size() == sizes.size() && offsets.size() == strides.size() &&
+           shape.size() == offsets.size());
+    return offsets.size();
+  }
+
+  bool isEmpty() const { return getRank() == 0 && !source && !scalar; }
+
+  // Process addition of two PtrStates.
+  LogicalResult addState(const PtrState &lhsState, const PtrState &rhsState,
+                         Operation *op, OpBuilder &builder) {
+    assert(isEmpty() && lhsState.getRank() == rhsState.getRank());
+    auto loc = op->getLoc();
+
+    if (lhsState.source && rhsState.source) {
+      op->emitRemark("TritonRaiseBlockPointer: do not support adding two "
+                     "pointer states that both have base pointers");
+      return failure();
+    }
+
+    source = lhsState.source ? lhsState.source : rhsState.source;
+
+    ArithBuilder abuilder(builder, loc);
+    for (uint64_t i = 0; i < lhsState.getRank(); i++) {
```

Comment:
Please change all postincrement to preincrement.
```suggestion
    for (uint64_t i = 0; i < lhsState.getRank(); ++i) {
```